### PR TITLE
Enhance perfomance by reducing icon updates

### PR DESCRIPTION
**Pull request by bjoern1001001**
I'm using ipytree in a project for displaying a hierarchy of multiple hundreds of objects. I found it very useful in this context, but noticed a considerable delay (> 30 sec) and high CPU and RAM (> 1 GB) usage when first constructing such large trees. So I looked into the code and found that triggerIconsUpdate() is called very often causing some overhead.

This pull request implements changes that greatly reduce the number of times icons are updated, thus also reducing time (down to ~5 sec for my use case) and CPU/RAM usage for building a large tree. If I did not miss anything, the layout of the tree remains consistent, so the icons are always shown as expected.

This is done by updating (open/close) icons only at the following times:
- When a child is added, the parent will update its icon.
- When a parent is opened, the icons of all children are updated recursively. This means if a node is opened, all its immediate children will update their icons and check if they are open. If so, their children will also update their icons and check if they are open, and so on. This makes sure all visible nodes have correct icons.
- Whenever the top level nodes are changed, icons are updated on all top level nodes and recursively on their visible children. This was introduced because adding a new top level node caused all icons to disappear.

The following is a relatively minimal example to showcase what I mean:

```
from ipytree import Tree, Node
tree = Tree(stripes=True)
tree
```

```
for i in range(5):
    tree.add_node(Node(str(i)))
    tree.nodes[i].opened = False
    for j in range(200):
        tree.nodes[i].add_node(Node(str(j)))
```

The time for running this code was reduced from about 30 seconds to below 5 seconds on my machine.

### CODE REVIEW

**Comment by martinRenou** [comment]
Thanks a lot for the PR, I will look into it :)

File: js/lib/tree.js [hunk 1]
```diff
@@ -106,12 +99,11 @@ var NodeView = widgets.WidgetView.extend({
             .find('i.jstree-icon.jstree-ocl').first();
     },
 
-    setOpenCloseIcon: function() {
-        var open_icon = this.getOpenIcon();
+    setOpenCloseIcon: function(recursive=false) {
+				var open_icon = this.getOpenIcon();
```
Comment (martinRenou) [inline js/lib/tree.js]: Do you mind replacing tabs by spaces everywhere? So we don't have a mix of tabs and spaces

File: js/lib/tree.js [hunk 2]
```diff
@@ -106,12 +99,11 @@ var NodeView = widgets.WidgetView.extend({
             .find('i.jstree-icon.jstree-ocl').first();
     },
 
-    setOpenCloseIcon: function() {
-        var open_icon = this.getOpenIcon();
+    setOpenCloseIcon: function(recursive=false) {
```
Comment (martinRenou) [inline js/lib/tree.js]: You added a `recursive` argument but don't seem to use it in the function (only passing it to `setOpenCloseIcon` on children)

File: js/lib/tree.js [hunk 3]
```diff
@@ -36,7 +31,6 @@ var NodeModel = widgets.WidgetModel.extend({
 
     initialize: function() {
         NodeModel.__super__.initialize.apply(this, arguments);
-
```
Comment (martinRenou) [inline js/lib/tree.js]: Nitpicking here, but do you mind not removing those empty lines? IMO it improves readability.

File: js/lib/tree.js (hunk 2 above)
Comment (martinRenou) [reply js/lib/tree.js]: Thanks a lot for updating the PR!! Could you please remove the `recursive` argument? Then I'll be really happy to merge it :)

Comment (martinRenou) [reply js/lib/tree.js]: Oh sorry, I see it is now used. LGTM

Comment (bjoern1001001) [reply js/lib/tree.js]: Thank you for looking at it so quickly.
I was actually planning on using the argument, I had forgotten to add the if statement (see l. 125) checking for it. I think in some cases the recursion does not make sense, e.g. if selectively changing the icon of the parent node. That's why the default is false, so `change:close_icon` etc. will not do a recursion.

**Comment by bjoern1001001** [comment]
Thank you for looking at my code and the helpful comments. Over the course of all the changes I had done I hadn't even noticed the formatting issues. I hope my code works as expected and I did not miss anything while checking this.

**Comment by martinRenou** [comment]
Thanks a lot for your PR, I tested locally and it really improved performances!

**Comment by martinRenou** [comment]
I wanted to squash the commits and merge, while keeping you as an author of the commit. Because I was not sure if Github could do that with the "squash and merge" button, I did it by hand. Your commit is now on master d71dd7212c2

**Pull request closed without merging**